### PR TITLE
Change: Towns don't build dead-end road bridges

### PR DESCRIPTION
**Pull request by 2TallTyler**
Currently, towns build bridges over ocean tiles or tracks without checking if the tile past the bridge will allow them to continue building the road. This frequently leads to bridges dead-ending at industries, houses, tracks, etc.

I have created a test savegame to demonstrate each case where the bridge should be allowed, and each case where it should be prohibited. You can run it on fast-forward to let the towns grow (I have included buses to drive growth). You can tell that a bridge has been rejected when the town grows the road to the end of the tile without bridging.
 
[Bridge_Test_Save.zip](https://github.com/OpenTTD/OpenTTD/files/5716592/Bridge_Test_Save.zip)

TrueBrain suggested using a DoCommand to try building a road in lieu of the industry/object/house/railway tile checks, but after some experimentation I decided that separate checks rejecting the bridge is more maintainable than one big `if` chain with a single `return false` at the end. Feel free to correct me, as I'm pretty new to all of this. :)

### CODE REVIEW

**Comment by 2TallTyler** [comment]
I can't see the whitespace which is failing CI in Visual Studio. Any tips?

**Comment by Milek7** [comment]
You need to squash these commits.

File: src/town_cmd.cpp [hunk 1]
```diff
@@ -1113,6 +1113,7 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 
 	/* We are in the right direction */
 	uint8 bridge_length = 0;      // This value stores the length of the possible bridge
+	TileIndex tile_past_bridge = tile; // We check the tile beyond the other end of the bridge to avoid building a dead-end bridge
```
Comment (Milek7) [inline src/town_cmd.cpp]: Initialization to tile is superfluous, as it is set below anyway.

File: src/town_cmd.cpp [hunk 2]
```diff
@@ -1136,9 +1137,52 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
 		} while (IsValidTile(bridge_tile) && (IsWaterTile(bridge_tile) || IsPlainRailTile(bridge_tile) || (IsNormalRoadTile(bridge_tile) && GetDisallowedRoadDirections(bridge_tile) != DRD_NONE)));
 	}
 
-	/* no water tiles in between? */
+	/* Don't allow bridges with no span between end tiles. */
 	if (bridge_length == 1) return false;
 
+	/* Next we check the tile beyond the bridge to ensure the road can continue. First check if the tile is valid. */
+	if (IsValidTile(bridge_tile)) {
```
Comment (Milek7) [inline src/town_cmd.cpp]: It shouldn't check for `bridge_tile + delta` validity instead?

**Comment by Milek7** [comment]
> TrueBrain suggested using a DoCommand to try building a road in lieu of the industry/object/house/railway tile checks, but after some experimentation I decided that separate checks rejecting the bridge is more maintainable than one big `if` chain with a single `return false` at the end.

I don't quite understand that, why it would result in big `if` chain?
Couldn't you just do `if (!IsRoadAllowedHere(t, tile_past_bridge, bridge_dir)) return false;`? (or maybe other conditions than `IsRoadAllowedHere` does, but with simple `DoCommand` instead of trying various situations)

File: src/town_cmd.cpp (hunk 2 above)
Comment (2TallTyler) [reply src/town_cmd.cpp]: Yes, thanks.

File: src/town_cmd.cpp (hunk 1 above)
Comment (2TallTyler) [reply src/town_cmd.cpp]: Right. Thanks.

**Comment by 2TallTyler** [comment]
The problem with returning false if no road is allowed to be built, is that there are several conditions in which the bridge should be allowed and returning false short-circuits those next checks. Specific situations are when a valid road tile already exists in the next tile: a road station, depot, bridge, tunnel, or a complete road tile facing the correct direction.

File: src/town_cmd.cpp [hunk 3]
```diff
@@ -1136,9 +1137,52 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
...
 
-	/* no water tiles in between? */
+	/* Don't allow bridges with no span between end tiles. */
 	if (bridge_length == 1) return false;
 
+	/* Next we check the tile beyond the bridge to ensure the road can continue. First check if the tile is valid. */
+	if (IsValidTile(bridge_tile + delta)) {
+		tile_past_bridge = bridge_tile + delta;
+	} else {
+		return false;
```
Comment (michicc) [inline src/town_cmd.cpp]: Personal opinion: swap the condition and then move `tile_past_bridge` out of the if.

File: src/town_cmd.cpp [hunk 4]
```diff
@@ -1136,9 +1137,52 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
...
+	if (IsTileType(tile_past_bridge, MP_TUNNELBRIDGE)) {
+		if (GetTunnelBridgeTransportType(tile_past_bridge) != TRANSPORT_ROAD || GetTunnelBridgeDirection(tile_past_bridge) != bridge_dir) return false;
+	}
+
+	/* If the next tile is a station, check if it's a road station facing the proper direction, otherwise disallow. Note that airport hangers are station tiles, so this catches them. */
+	if (IsTileType(tile_past_bridge, MP_STATION)) {
+		/* If not a road station, disallow. */
+		if (!IsRoadStop(tile_past_bridge)) return false;
+		/* If the next tile is a road station, allow if it's facing the same direction, otherwise disallow. */
+		if (IsRoadStop(tile_past_bridge)) {
```
Comment (michicc) [inline src/town_cmd.cpp]: Somewhat redundant with the line above.

File: src/town_cmd.cpp (hunk 3 above)
Comment (2TallTyler) [reply src/town_cmd.cpp]: Agreed.

**Comment by TrueBrain** [comment]
I also send you some words on Discord too, but in summary here too:

You now define all the possible ways the bridge would be invalid, which is a pretty lengthy list. Possibly it is better to do the reverse: define what are valid scenarios, as I think that list would be a lot shorter:

- Road already connecting
- Can build road that connects
- Station right orientation
- Depot right orientation
- Other tunnel or bridge in right orientation

C'est tout!

This would be, I think, much easier code to read, but also much less likely to have a flaw somewhere in the logic :)

File: src/town_cmd.cpp (hunk 4 above)
Comment (2TallTyler) [reply src/town_cmd.cpp]: Indeed. Thanks.

**Comment by 2TallTyler** [comment]
Per my conversation with TrueBrain on Discord, I will be rewriting this to use a new function. Stay tuned.

File: src/town_cmd.cpp [hunk 5]
```diff
@@ -1087,6 +1087,59 @@ static bool GrowTownWithRoad(const Town *t, TileIndex tile, RoadBits rcmd)
...
+static bool CanContinueRoad(const TileIndex end_tile, const DiagDirection bridge_dir)
+{
+	const int delta = TileOffsByDiagDir(bridge_dir); // +1 tile in the direction of the bridge
+	TileIndex next_tile = end_tile + delta; // The tile beyond the bridge
+
+	/* Before we try anything, make sure the tile is on the map and not the void. */
+	if (!IsValidTile(end_tile + delta)) return false;
+
+	/* If the tile is clear or has trees, it is fine to build. */
+	if (IsTileType(next_tile, MP_CLEAR) || IsTileType(next_tile, MP_TREES)) return true;
```
Comment (TrueBrain) [inline src/town_cmd.cpp]: This is a lie! Sloping can prevent building :) Here is where the `IsRoadAllowedHere` comes into play. I think you can change this if, and the next 2, with only a call to this function. I might be wrong :)

I would do:

- TunnelBridge check
- Station check
- `IsRoadAllowedHere` call

Comment (2TallTyler) [reply src/town_cmd.cpp]: `IsRoadAllowedHere` didn't work, but I used a `DoCommand` like your originally suggested for the same purpose.

**Comment by 2TallTyler** [comment]
I need to sleep before I break anything further. I'll try this again in the morning.

The correct commit should be https://github.com/OpenTTD/OpenTTD/commit/cc53efa41220b1e661d996bb27808939829dd592

**Comment by 2TallTyler** [comment]
Per my discussion with TrueBrain on Discord, I appear to have broken GitHub and the easiest fix is starting a new branch and making a new PR.

**Pull request closed without merging**